### ISSUE TITLE
False positive `overload-overlap` with booleans and integer literals

### ISSUE THREAD

**Issue by bzoracler** [python/mypy#20084]
**Bug Report, To Reproduce, & Actual Behaviour**

The following 2 cases trigger `overload-overlap`, when neither of them should (see [mypy Playground](https://mypy-play.net/?mypy=latest&python=3.13&gist=9ed93bc23c8183730ff3fabad36d724a)).

```python
from typing import overload, Literal

@overload
def display_str_literal(val: Literal[False], /) -> Literal["BooleanFalse"]: ...  # E: ... overload-overlap
@overload
def display_str_literal(val: Literal[0], /) -> Literal["Zero"]: ...
```

```python
from typing import overload, Literal

@overload
def display_str_literal_2(val: bool, /) -> Literal["Boolean"]: ...
@overload
def display_str_literal_2(val: Literal[-1], /) -> Literal["NegativeOne"]: ...
```

- Mypy version used: 1.18.2
- Python version used: 3.12, 3.13

**Comment by wyattscarpenter** [comment]
As a workaround, the second one will work if you use `Literal[True, False]` instead of `bool` (related to https://github.com/python/mypy/issues/14764).

**Comment by jorenham** [comment]
> As a workaround, the second one will work if you use `Literal[True, False]` instead of `bool` (related to [#14764](https://github.com/python/mypy/issues/14764)).

Note that ruff (pyupgrade if I remember correctly) can report this as an error (depending on your config, of course). Using literals also tends to negatively affect type-checker analysis times (although it's probably negligible in most cases). So I think that in this case I'd prefer using a classic `# type: ignore[overload-overlap]` as workaround.
